### PR TITLE
fix the bug: memory size is twice as much as block_size

### PR DESCRIPTION
**Pull request by kr11**
Block_size is the threshold of flushing disk and releasing resources. 
We write one device and one sensor using Plain encoding and set block_size is 32M and page_size is 4k.
Although -Xmx is set to 50M, it will throw java heap space error.
This pr fix this bug.

### CODE REVIEW

File: src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
+package cn.edu.thu.tsfile.common.utils;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/***
+ * BytesInput is a abstract class encapsulating several operations related to OutputStream with
```
Comment (jixuan1989) [inline src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java]: The javadoc is stale

File: src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    public ListByteArrayOutputStream(ByteArrayOutputStream ... param) {
+        list = new ArrayList<>();
+        for (ByteArrayOutputStream out : param){
+            list.add(out);
+            totalSize += out.size();
+        }
+    }
+
+    /**
+     * input an OutputStream as parameter. BytesInput output its data to this OutputStream
```
Comment (jixuan1989) [inline src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java]: Write all the data into the given OutputStream.

File: src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java (hunk 1 above)
Comment (kr11) [inline src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java]: I have update the javadoc in this class and will commit it later.

File: src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java (hunk 2 above)
Comment (kr11) [inline src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java]: I have update the javadoc in this class and will commit it later.

Comment (kr11) [inline src/main/java/cn/edu/thu/tsfile/common/utils/ListByteArrayOutputStream.java]: a new commit has been pushed

File: src/main/java/cn/edu/thu/tsfile/timeseries/write/io/TSFileIOWriter.java [hunk 3]
```diff
@@ -72,13 +65,13 @@ public TSFileIOWriter(FileSchema schema,TSRandomAccessFileWriter output, long of
     }
 
     /**
-     * write given BytesInput to output stream. This method is called when total memory size exceeds
-     * the row group size threshold.
+     * Writes given <code>ListByteArrayOutputStream</code> to output stream.
+     * This method is called when total memory size exceeds the row group size threshold.
      * 
      * @param bytes - data in page writer
```
Comment (qiaojialin) [inline src/main/java/cn/edu/thu/tsfile/timeseries/write/io/TSFileIOWriter.java]: data in all page writers

File: src/test/java/cn/edu/thu/tsfile/common/utils/ListByteOutputStreamTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
+package cn.edu.thu.tsfile.common.utils;
+
+import cn.edu.thu.tsfile.common.utils.ListByteArrayOutputStream;
```
Comment (qiaojialin) [inline src/test/java/cn/edu/thu/tsfile/common/utils/ListByteOutputStreamTest.java]: Unused import

File: src/test/java/cn/edu/thu/tsfile/compress/CompressTest.java [hunk 5]
```diff
@@ -37,7 +36,7 @@ public void noCompressorTest() throws UnsupportedEncodingException, IOException
       out.write(inputString.getBytes("UTF-8"));
```
Comment (qiaojialin) [inline src/test/java/cn/edu/thu/tsfile/compress/CompressTest.java]: noCompressorTest method does not throw UnsupportedEncodingException, and all "compresser"s of inputString should be "compressor"?

File: src/test/java/cn/edu/thu/tsfile/compress/CompressTest.java [hunk 6]
```diff
@@ -49,7 +48,7 @@ public void snappyCompressorTest() throws UnsupportedEncodingException, IOExcept
...
```
Comment (qiaojialin) [inline src/test/java/cn/edu/thu/tsfile/compress/CompressTest.java]: UnsupportedEncodingException can removed from all 3 test methods

File: src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java [hunk 7]
```diff
@@ -33,7 +32,7 @@ public void testNoFreq() {
         short s1 = 12;
         boolean b1 = false;
         int i1 = 1;
-        long l1 = 123142120391l;
+        long l1 = 123142120391L;
         float f1 = 2.2f;
         double d1 = 1294283.4323d;
         // BigDecimal big1 = new BigDecimal("123.43");
```
Comment (qiaojialin) [inline src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java]: remove commented code

File: src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java [hunk 8]
```diff
@@ -49,13 +48,11 @@ public void testNoFreq() {
             // writer.write(timeCount++, big1);
```
Comment (qiaojialin) [inline src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java]: remove commented code

File: src/main/java/cn/edu/thu/tsfile/timeseries/write/io/TSFileIOWriter.java (hunk 3 above)
Comment (qiaojialin) [inline src/main/java/cn/edu/thu/tsfile/timeseries/write/io/TSFileIOWriter.java]: data in all page writers

File: src/main/java/cn/edu/thu/tsfile/timeseries/write/page/PageWriterImpl.java [hunk 9]
```diff
@@ -113,15 +100,18 @@ public void writeToFileWriter(TSFileIOWriter writer, Statistics<?> statistics)
...
         // return size of buffer + page max size;
-        int digestSize = (totalValueCount==0) ? 0 : desc.getTypeLength() * 2;
-        return buf.size()
-                + TSFileIOWriter.metadataConverter.caculatePageHeaderSize(digestSize);
+        return buf.size() + estimateMaxPageHeaderSize();
+    }
+
+    private int estimateMaxPageHeaderSize() {
+        int digestSize = (totalValueCount == 0) ? 0 : desc.getTypeLength() * 2;
+        return TSFileIOWriter.metadataConverter.caculatePageHeaderSize(digestSize);
```
Comment (kr11) [inline src/main/java/cn/edu/thu/tsfile/timeseries/write/page/PageWriterImpl.java]: fixed, thanks

File: src/test/java/cn/edu/thu/tsfile/common/utils/ListByteOutputStreamTest.java (hunk 4 above)
Comment (kr11) [inline src/test/java/cn/edu/thu/tsfile/common/utils/ListByteOutputStreamTest.java]: fixed, thanks

File: src/test/java/cn/edu/thu/tsfile/compress/CompressTest.java (hunk 5 above)
Comment (kr11) [inline src/test/java/cn/edu/thu/tsfile/compress/CompressTest.java]: "compresser" and unused exception has been fixed

File: src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java (hunk 7 above)
Comment (kr11) [inline src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java]: fixed, thx

File: src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java (hunk 8 above)
Comment (kr11) [inline src/test/java/cn/edu/thu/tsfile/timeseries/write/series/ValueWriterTest.java]: fixed, thx

File: src/main/java/cn/edu/thu/tsfile/timeseries/write/io/TSFileIOWriter.java (hunk 3 above)
Comment (kr11) [inline src/main/java/cn/edu/thu/tsfile/timeseries/write/io/TSFileIOWriter.java]: this annotation has been changed

File: src/main/java/cn/edu/thu/tsfile/timeseries/write/InternalRecordWriter.java [hunk 10]
```diff
@@ -123,7 +123,7 @@ protected void checkMemorySize() throws IOException {
 				flushRowGroup(true);
 				recordCountForNextMemCheck = rowGroupSizeThreshold / oneRowMaxSize;
```
Comment (qiaojialin) [inline src/main/java/cn/edu/thu/tsfile/timeseries/write/InternalRecordWriter.java]: checkMemorySize method in this class should add a constraint on record count. Flush rowgroup When record count reaches the threshold

Comment (kr11) [inline src/main/java/cn/edu/thu/tsfile/timeseries/write/InternalRecordWriter.java]: has been added

**Comment by qiaojialin** [comment]
+1

**Pull request merged**